src/codegen/query_cache.cpp

```diff
@@ -0,0 +1,78 @@
+//===----------------------------------------------------------------------===//
```

Comment:
There seems to be a design issue in this PR.

The k-v store in the cache is a `std::map`, which requires a user-defined "less than" comparator, instead of "not equal to".

The reason is that what's inside `std::map` is a balanced binary search tree. It needs a "less than" comparator to perform O(log n) operations. When it wants to determine whether two keys are equal, it just does "!(A < B) && !(B < A)".

The reason why the test cases are passing is that they only test situations when the cache has at most 1 element. But if we test situations where we have more keys, things can go weired.

Consider this example.

```C++
#include <map>
#include <iostream>

using namespace std;

struct Cmp {
    bool operator()(const int &lhs, const int &rhs) const {
        return lhs != rhs;
    }
};

int main() {
    map<int, int, Cmp> m;
    int n = 10;
    for (int i = 0; i < n; ++i) {
        m.emplace(i, i);
    }

    int cnt = 0;
    for (int i = 0; i < n; ++i) {
        cnt += m.count(i);
    }
    std::cout << m.size() << std::endl; // 10
    std::cout << cnt << std::endl; // 5 ???
    return 0;
}
```

If we are to use `std::unordered_map` instead of `std::map`, we would need a hash function as well as equality comparison.

Reply:
I can confirm :-). I admit that I was careless about this. A good catch.